components/sup/src/manager/spec_watcher.rs

```diff
@@ -12,659 +12,298 @@
-                            watch_path.display(),
-                            err
-                        );
-                        return;
-                    }
-                };
-                if let Err(err) = watcher.watch(&watch_path, RecursiveMode::NonRecursive) {
-                    outputln!(
-                        "SpecWatcher({}) could not start fs watching, ending thread ({})",
-                        watch_path.display(),
-                        err
-                    );
-                    return;
-                }
-                while let Ok(event) = rx.recv() {
-                    debug!(
-                        "SpecWatcher({}) file system event: {:?}",
-                        watch_path.display(),
-                        event
-                    );
-                    have_events.store(true, Ordering::Relaxed);
-                }
-                outputln!(
-                    "SpecWatcher({}) fs watching died, restarting thread",
-                    watch_path.display()
-                );
-                drop(watcher);
-                Self::setup_watcher::<W>(watch_path.clone(), have_events.clone()).unwrap();
-            })?;
-        Ok(())
-    }
+/// Provides an abstraction layer over filesystem notifications for
+/// spec files.
+pub struct SpecWatcher {
+    // Not actually used; only holding onto it for lifetime / Drop
+    // purposes (`Drop` kills the threads that the watcher spawns to do
+    // its work).
+    _watcher: RecommendedWatcher,
+    channel: Receiver<DebouncedEvent>,
+}
 
-    fn have_fs_events(&mut self) -> bool {
-        self.have_events.load(Ordering::Relaxed)
+impl SpecWatcher {
+    /// Start up a separate thread to listen for filesystem
+    /// events.
+    pub fn run(spec_dir: &SpecDir) -> Result<SpecWatcher> {
+        // The act of creating a `notify::Watcher` creates threads on
+        // its own. It does not, however, allow you to set the _names_
+        // of those threads.
+        //
+        // We're creating a SpecWatcher in a thread just so we can get
+        // some control over the name of the threads that the
+        // underlying `notify::Watcher` creates (_that_ is what this
+        // function's documentation is referring to), which makes
+        // monitoring and reasoning about the overall Supervisor
+        // process easier. There's no other reason than that; if the
+        // `notify` crate allowed us to name the threads, we could
+        // just use a slightly modified version of `SpecWatcher::new`
+        // instead.
+
+        // I'd rather not have to do this clone, but it's a side
+        // effect of this thread business. Better to eat it here
+        // rather than at the callsite; having this function take a
+        // reference is the true API we want.
+        let dir = spec_dir.clone();
+
+        Builder::new()
+            .name(String::from("spec-watcher"))
+            .spawn(move || Self::new(&dir))?
```

Comment:
My reading of the docs is that `spawn` will return an error if it couldn't spawn the thread for some OS-related reason (and not anything related to the code the thread is running). I think we still want our `run` function to bubble that error up, rather than causing a panic, which `expect` would do.

Calling `join` will then return the final value of the thread, or an error if it panicked. The `map_err` call handles the thread panic scenario, whereas the final `?` at the _very_ end unwraps the result of calling `SpecWatcher::new`, which could cause an error itself. 

I have to return the pure `Result` from the thread because you can't call `?` in the thread.

Reply:
Oh, I should've looked at the docs for [`std::thread::spawn`](http://doc.rust-lang.org/1.30.1/std/thread/fn.spawn.html) and not just `Builder::spawn`. You're quite correct, and this code seems just right.